### ISSUE TITLE
Reorganize doesn't execute successfully after 10.2.7 update.

### ISSUE THREAD

**Issue by adamjstone** [codecadwallader/codemaid#382]
### Environment
- Visual Studio version: 2015 Community
- CodeMaid version: example 10.2.7
- Code language: C#

### Description
After updating to 10.2.7 reorganize doesn't execute successfully for manual or automatic (on save) cleanup.

### Steps to recreate
1. Using the latest production release for VS 2015 Community, open an organized C# class file.
2. Cut and paste a class member to a known out-of-order location within the file.
3. Execute cleanup and observe that the file is not reorganized.

### Current behavior
The reorganize step does not execute successfully during cleanup.

### Expected behavior
Reorganize functionality should be restored.

**Comment by adamjstone** [comment]
Configuration attached.

[Issue 382 example.zip](https://github.com/codecadwallader/codemaid/files/680150/Issue.382.example.zip)

**Comment by codecadwallader** [comment]
Thanks for reporting. Can you enable diagnostics mode
(CodeMaid->Options->General->Diagnostics mode checkbox) and after repeating
the steps check the Output tool window to see what (if any) error is being
thrown by reorganization?

**Comment by adamjstone** [comment]
Howdy. Doesn't appear that any exception is being raised.

[Output-CodeMaid.txt](https://github.com/codecadwallader/codemaid/files/680288/Output-CodeMaid.txt)

**Comment by codecadwallader** [comment]
Thanks for the log, and it does appear that reorganization is running.  I haven't been able to reproduce it with VS2017, so here's a few things to try:

1.  Create a new solution with a new C# class library.  Within the generated Class1.cs add two fields "field2" and "field1".  Directly invoke CodeMaid's invocation (vs. indirectly through cleaning).  Do you still see the issue?
2.  Reset your settings to default, and attempt the previous operation again.  This will rule out any setting combinations that may be causing a problem.  You can re-import your settings provided above afterwards.
3.  You mentioned after updating to the latest 10.2.7.  I'm assuming you were previously using the 10.1 series?
4.  If you look in the Spade tool window, is the file being represented correctly?  This is a visual representation of the code model that is used for reorganization.
5.  If you use Spade to drag and drop members, do they move successfully?  Under the covers this uses the same code movement logic as reorganization.

I'm hoping one of those points us in a direction to what may be causing the issue.

**Comment by adamjstone** [comment]
Results:

1. It worked! Cleanup on save produces the desired results, too, using the solution configuration from the failing scenario. It has something to do with my solution or my files, I presume.
2. Skipping this step.
3. Blargh. I'm not sure which version I was using previously because my extensions auto-update. I observed the problem as soon as I started working yesterday and noticed that Visual Studio updated CodeMaid earlier in the morning.
4. Roger that. It's correct.
5. They do. Cleanup on save and manual invocation of reorganize still doesn't cause members to reorganize, though. The other features are working on cleanup, however.

### Aha!

After playing with my solution for a while, I found the problem. It has something to do with the particulars of the attribute below. If I remove or replace `StructLayout` with a different class attribute, reorganization works. What do you make of it?

```
/// <summary>
/// Represents a quantitative measurement of three-dimensional space.
/// </summary>
[StructLayout(LayoutKind.Explicit, CharSet = CharSet.Unicode, Size = 16)]
public struct Volume : IBinaryDecomposable, IComparable<Volume>, IEquatable<Volume>
```

**Comment by adamjstone** [comment]
Thinking about this further and, without looking at the CodeMaid source, my prediction is that the observed behavior is intentional (to prevent screwing up sequential struct layouts). Am I right?

Maybe I should resubmit this as an enhancement request. Something like: "Do not suppress reorganization for explicit struct layout." Thoughts?

**Comment by codecadwallader** [comment]
Yes, we explicitly look for StructLayoutAttribute and exclude those members from reorganization.  I believe it's in the CodeReorganizationAvailabilityLogic class but a search would confirm.  It's been there for awhile so we don't do rearrange something that shouldn't.  I've never used the attribute myself, are there different modes and some should still allow reorganization?

**Comment by adamjstone** [comment]
In my view, yes. When `LayoutKind.Sequential` is specified the bits are ordered in memory as they are specified, sequentially, in the file. Therefore, alphabetizing them could produce a behavioral change. Generally, when defining a struct, I use `LayoutKind.Explicit` with `FieldOffsetAttribute` decorations to prevent such shenanigans because, like most CodeMaid users, I prefer file organization to be uniform and rule-based.

**Comment by adamjstone** [comment]
I'll take it on. It looks like the change should go near line 226 of  CodeReorganizationManager .

Thanks, sir.

**Comment by codecadwallader** [comment]
Wonderful, thanks!  You may find some helpful notes at http://www.codemaid.net/contribute/
